core/rawdb/prunedfreezer.go

```diff
@@ -0,0 +1,286 @@
+)
+
+// nodatafreezer is an empty freezer, only record 'frozen' , the next recycle block number form kvstore.
+type prunedfreezer struct {
+	db ethdb.KeyValueStore // Meta database
+	// WARNING: The `frozen` field is accessed atomically. On 32 bit platforms, only
+	// 64-bit aligned fields can be atomic. The struct is guaranteed to be so aligned,
+	// so take advantage of that (https://golang.org/pkg/sync/atomic/#pkg-note-BUG).
+	frozen    uint64 // BlockNumber of next frozen block
+	threshold uint64 // Number of recent blocks not to freeze (params.FullImmutabilityThreshold apart from tests)
+
+	instanceLock fileutil.Releaser // File-system lock to prevent double opens
+	quit         chan struct{}
+	closeOnce    sync.Once
+}
+
+// newNoDataFreezer creates a chain freezer that deletes data enough ‘old’.
+func newPrunedFreezer(datadir string, db ethdb.KeyValueStore) (*prunedfreezer, error) {
+	if info, err := os.Lstat(datadir); !os.IsNotExist(err) {
+		if info.Mode()&os.ModeSymlink != 0 {
+			log.Warn("Symbolic link ancient database is not supported", "path", datadir)
+			return nil, errSymlinkDatadir
+		}
+	}
+
+	lock, _, err := fileutil.Flock(filepath.Join(datadir, "../NODATA_ANCIENT_FLOCK"))
+	if err != nil {
+		return nil, err
+	}
+
+	freezer := &prunedfreezer{
+		db:           db,
+		threshold:    params.FullImmutabilityThreshold,
+		instanceLock: lock,
+		quit:         make(chan struct{}),
+	}
+
+	if err := freezer.repair(datadir); err != nil {
+		return nil, err
+	}
+
+	log.Info("Opened ancientdb with nodata mode", "database", datadir, "frozen", freezer.frozen)
+	return freezer, nil
+}
+
+// repair init frozen , compatible disk-ancientdb and pruner-block-tool.
+func (f *prunedfreezer) repair(datadir string) error {
+	// compatible prune-block-tool
+	offset := ReadOffSetOfCurrentAncientFreezer(f.db)
+
+	// compatible freezer
+	min := uint64(math.MaxUint64)
+	for name, disableSnappy := range FreezerNoSnappy {
+		table, err := NewFreezerTable(datadir, name, disableSnappy)
+		if err != nil {
+			return err
+		}
+		items := atomic.LoadUint64(&table.items)
+		if min > items {
+			min = items
+		}
+		table.Close()
+	}
+	offset += min
+
+	if frozen := ReadFrozenOfAncientFreezer(f.db); frozen > offset {
+		offset = frozen
+	}
+
+	atomic.StoreUint64(&f.frozen, offset)
+	if err := f.Sync(); err != nil {
+		return nil
+	}
+	return nil
+}
+
+// Close terminates the chain prunedfreezer.
+func (f *prunedfreezer) Close() error {
+	var err error
+	f.closeOnce.Do(func() {
+		close(f.quit)
+		f.Sync()
+		err = f.instanceLock.Release()
+	})
+	return err
+}
+
+// HasAncient returns an indicator whether the specified ancient data exists, return nil.
+func (f *prunedfreezer) HasAncient(kind string, number uint64) (bool, error) {
+	return false, nil
+}
+
+// Ancient retrieves an ancient binary blob from prunedfreezer, return nil.
+func (f *prunedfreezer) Ancient(kind string, number uint64) ([]byte, error) {
+	if _, ok := FreezerNoSnappy[kind]; ok {
```

Comment:
It will always returnn nil, try simplify it?

Reply:
I think the type of error should be distinguished. For tables outside the anti-counterfeiting range or other non-existent tables, the corresponding error should be returned. The upper layer can respond according to different situations. Just returning nil may affect the judgment of the upper layer.
At the same time, it can also prevent simple bugs. For example, when the tool accesses ancient, the name of the table may be misspelled, and it is impossible to locate the problem by simply returning nil.